src/miner_lora.erl

```diff
@@ -381,7 +381,11 @@ handle_udp_packet(<<?PROTOCOL_2:8/integer-unsigned,
     Gateway =
         case maps:find(MAC, Gateways) of
             {ok, #gateway{received=Received}=G} ->
-                G#gateway{ip=IP, port=Port, received=Received+1};
+                %% We purposely do not update gateway's port here. It
+                %% should only be updated when handling PULL_DATA,
+                %% otherwise we may send downlink packets to the wrong
+                %% port.
+                G#gateway{ip=IP, received=Received+1};
```

Comment:
Maybe don't update IP here as well then?

Reply:
Huh, I didn't think about that.